### DISCUSSION TITLE
Initializing state from external data

### DISCUSSION THREAD

**Discussion by ryaninvents** [statelyai/xstate#2865] [Q&A] [answered]
My intended use is for handling customer subscription lifecycle events. We don't host any of the state ourselves -- we need to check in with third parties for information such as number of active subscriptions and subscription renewal date.

I want to use XState as part of our backend. The logic has gotten kind of unwieldy, and I really want the clarity and debuggability that XState can bring to the party. (Plus, it would be super cool to be able to display a diagram in our admin dashboard for a given customer.) Here's [a partial mockup](https://stately.ai/viz/c08b9398-56a2-4eb2-a007-d90f4b03458a) of the way I want to build this.

<a href="https://stately.ai/viz/c08b9398-56a2-4eb2-a007-d90f4b03458a"><img alt="20211207 142801 Google Chrome screenshot" src="https://user-images.githubusercontent.com/8356669/144966359-83c1e403-3fad-4bf7-8d0c-c2276ffbe282.png"></a>

However, I'm working to interpret [this line](https://xstate.js.org/docs/guides/states.html#:~:text=You%20should%20never%20have%20to%20create%20a%20State%20instance%20manually) of the docs:

>   You should never have to create a `State` instance manually. Treat `State` as a read-only object that *only* comes from `machine.transition(...)` or `service.onTransition(...)`.

_Immediately_ above this text is a call `fetchMachine.transition('loading', /* ... */)`. To me, there is no conceptual difference between the string `'loading'` and a hand-built more complex state object such as `{Subscriber: {Questionnaire: 'Incomplete'}}` if they're both constructed from sources external to XState. The ["Persisting state" example](https://xstate.js.org/docs/guides/states.html#persisting-state) seems to treat state as a serializable opaque object, but it seems like a simple enough format that I'm ready to assume it will be stable in the long run.

I'm not planning to "create a `State` instance manually", but I _am_ planning to call `machine.transition()` with an object composed from external sources -- for example `machine.transition({Subscriber: {Questionnaire: 'Incomplete'}}, {type: 'CompleteQuestionnaire'})`. My question: **am I inviting disaster by pulling from external sources to form the argument to `machine.transition()`? Or is this a perfectly good way to use XState?**

I've been exploring this approach on RunKit [here](https://runkit.com/r24y/serializing-deserializing-xstate-machines) and [here](https://runkit.com/r24y/subscription-state-machine), and it seems solid. I just want assurance that it's okay to compose the argument to `machine.transition()` like this.

**Comment by davidkpiano** [comment, marked as answer]
The reason that warning is there is because it's possible to create a `State` that is incompatible with the machine; for instance, `machine.transition({ foo: 'impossible' }, ... )`.

If you can guarantee somehow that the state is compatible, then yes, you can construct the state yourself.

**Reply by ryaninvents** [reply]
@davidkpiano Thank you so much! I think this should be feasible then, and will save us a lot of effort.
